### PR TITLE
[SPARK-38292][PYTHON]Support na_filter for pyspark.pandas.read_csv

### PR DESCRIPTION
**Pull request by pralabhkumar**
### What changes were proposed in this pull request?
na filter is added in the read csv option . This is similar to na filter option in pandas

data.csv
```
A,B,C
,val1,val2
val3
```
```python
from pyspark import pandas as ps
import pandas as pd
ps.read_csv("data.csv")

      A     B     C
0  None  val1  val2
1  val3  None  None

pd.read_csv("data.csv")

      A     B     C
0   NaN  val1  val2
1  val3   NaN   NaN

ps.read_csv("data.csv", na_filter=False)

      A     B     C
0        val1  val2
1  val3            
  
pd.read_csv("data.csv", na_filter=False)

      A     B     C
0        val1  val2
1  val3            

```
### Why are the changes needed?
Added na_filter option

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
Unit test cases

### CODE REVIEW

**Comment by pralabhkumar** [comment]
@HyukjinKwon  Please review the PR

**Comment by pralabhkumar** [comment]
@HyukjinKwon  Gentle ping

**Comment by HyukjinKwon** [comment]
@itholic can you review this one please?

**Comment by itholic** [comment]
Could you refine the PR description ??

Seems like the format is broken in somewhere so it's a bit hard to understand the example and purpose.

**Comment by itholic** [comment]
And also fix the title ?? I think we can just use the JIRA title, "Support `na_filter` for pyspark.pandas.read_csv" as is.

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -550,6 +552,7 @@ def func(iterator):
             # There aren't any jvm api for creating a dataframe from rdd storing csv.
             # We can do it through creating a jvm dataset firstly and using the jvm api
             # for creating a dataframe from dataset storing csv.
+
```
Comment (itholic) [inline python/pyspark/sql/readwriter.py:555]: Let's revert the unrelated change ?

File: python/pyspark/pandas/namespace.py [hunk 2]
```diff
@@ -285,6 +286,9 @@ def read_csv(
         Indicates the encoding to read file
     options : dict
         All other options passed directly into Spark's data source.
+    na_filter : bool
+        If na_filter is false missing values will remain as is otherwise it
+        will be converted to None. By default it will True
```
[the remark this answers is not in the document]
Comment (pralabhkumar) [reply python/pyspark/pandas/namespace.py]: Have done the changes . However , in case of Spark IMO , there wouldn't be any performance improvement , since it is using Univocity parsers which handle the null  when reading data.  Therefore I have not added performance related things in the documenataion

File: python/pyspark/pandas/namespace.py [hunk 3]
```diff
@@ -285,6 +286,9 @@ def read_csv(
         Indicates the encoding to read file
     options : dict
         All other options passed directly into Spark's data source.
+    na_filter : bool
```
[the remark this answers is not in the document]
Comment (pralabhkumar) [reply python/pyspark/pandas/namespace.py]: done

File: python/pyspark/sql/readwriter.py (hunk 1 above)
Comment (pralabhkumar) [reply python/pyspark/sql/readwriter.py]: done

**Comment by pralabhkumar** [comment]
@itholic  Have done the changes as suggested

**Comment by pralabhkumar** [comment]
@itholic @HyukjinKwon Gentle ping

**Comment by pralabhkumar** [comment]
Gentle ping to please review the PR @itholic

**Comment by pralabhkumar** [comment]
@itholic  Please review the PR

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 4]
```diff
@@ -317,7 +317,15 @@ class UnivocityParser(
         if (skipRow) {
           row.setNullAt(i)
         } else {
-          row(i) = valueConverters(i).apply(getToken(tokens, i))
+          // This is required to not set value as null ,
+          // 1. If its missing value at the end of line.
+          // 2. If the missing value at the beginning of line.
```
Comment (itholic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:322]: nit: `1. If its missing value` ... -> `1. If "the" missing value` ?

**Review by itholic — COMMENTED** [review]
Otherwise seems fine to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala (hunk 4 above)
Comment (pralabhkumar) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: @itholic  Have done the changes as suggested .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 5]
```diff
@@ -317,7 +317,15 @@ class UnivocityParser(
         if (skipRow) {
           row.setNullAt(i)
         } else {
-          row(i) = valueConverters(i).apply(getToken(tokens, i))
+          // This is required to not set value as null ,
+          // 1. If the missing value at the end of line.
+          // 2. If the missing value at the beginning of line.
+          if (!options.naFilter && (i>=tokens.length ||
+            (i==0 && getToken(tokens, i).length == 0))) {
+            row(i) = valueConverters(i).apply("")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:325]: For `""`, it's interpreted differently with `nullValue` or `emptyValue`. I don't think we should rely on this conversion here. Can we control this by setting one of these options?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 6]
```diff
@@ -317,7 +317,15 @@ class UnivocityParser(
         if (skipRow) {
           row.setNullAt(i)
         } else {
-          row(i) = valueConverters(i).apply(getToken(tokens, i))
+          // This is required to not set value as null ,
+          // 1. If the missing value at the end of line.
+          // 2. If the missing value at the beginning of line.
+          if (!options.naFilter && (i>=tokens.length ||
+            (i==0 && getToken(tokens, i).length == 0))) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:324]:
```suggestion
            (i == 0 && getToken(tokens, i).length == 0))) {
```
should add a space between operator per the style guide

**Comment by HyukjinKwon** [comment]
Hey, I think the fix here is too hacky. Can we make this working independently with other options being set?

**Comment by pralabhkumar** [comment]
> Hey, I think the fix here is too hacky. Can we make this working independently with other options being set?

Hi @HyukjinKwon 

Thx for reviewing  . Had again gone through the code. Here is my understanding (same is mentioned in jira)

IMHO 
Setting nullValue option  will not help here . Since whatever the value we set ',,] string will be converted to the value  by Univocityparser(external)  which we have set. 
For e.g A,, and if setNullValue(“B”)  will result to A,B by univocity parser . Then spark univocity parser in nullSafeDatum
Will always convert to null (since datum == options.nullValue) .  So output will always be A, null whereas we need A,,
.There IMHO setting nullValue will not help here unless we have options.naFilter value to False which will make sure the above condition doesn't satisfy. 

Now in case of missing values in the beginning and end of line , current logic in convert method of UnivocietyParser is to go into exception
row.update(i, requiredSchema.existenceDefaultValues(i)) and update with default value.  Now we don’t want values to set to null in case options.naFilter is false.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala (hunk 5 above)
Comment (pralabhkumar) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:325]: This is only called if the missing values in the beginning or end . In that case current code goes into exception block and update row with default values (which will replace missing values with null) . In case option.naFilter is False , we do not want to replace with null and let them be missing value
